packages/cli/src/modules/mcp/mcp-oauth-token.service.ts

```diff
@@ -0,0 +1,210 @@
+	): Promise<void> {
+		await this.accessTokenRepository.manager.transaction(async (transactionManager) => {
+			await transactionManager.save(this.accessTokenRepository.target, {
+				token: accessToken,
+				clientId,
+				userId,
+			});
+
+			await transactionManager.save(this.refreshTokenRepository.target, {
+				token: refreshToken,
+				clientId,
+				userId,
+				expiresAt: Date.now() + this.REFRESH_TOKEN_EXPIRY_MS,
+			});
+		});
+	}
+
+	async validateAndRotateRefreshToken(
+		refreshToken: string,
+		clientId: string,
+	): Promise<OAuthTokens> {
+		const refreshTokenRecord = await this.refreshTokenRepository.findOne({
+			where: {
+				token: refreshToken,
+				clientId,
+			},
+		});
+
+		if (!refreshTokenRecord) {
+			throw new Error('Invalid refresh token');
+		}
+
+		if (refreshTokenRecord.expiresAt < Date.now()) {
+			await this.refreshTokenRepository.remove(refreshTokenRecord);
+			throw new Error('Refresh token expired');
+		}
+
+		const { accessToken, refreshToken: newRefreshToken } = this.generateTokenPair(
+			refreshTokenRecord.userId,
+			clientId,
+		);
+
+		await this.refreshTokenRepository.remove(refreshTokenRecord);
+
+		await this.saveTokenPair(accessToken, newRefreshToken, clientId, refreshTokenRecord.userId);
+
+		this.logger.info('Refresh token rotated and new access token issued', {
+			clientId,
+			userId: refreshTokenRecord.userId,
+		});
+
+		return {
+			access_token: accessToken,
+			token_type: 'Bearer',
+			expires_in: this.ACCESS_TOKEN_EXPIRY_SECONDS,
+			refresh_token: newRefreshToken,
+		};
+	}
+
+	async verifyAccessToken(token: string): Promise<AuthInfo> {
+		let decoded;
+
+		try {
+			decoded = this.jwtService.verify(token);
+		} catch (error) {
+			throw new Error('Invalid access token: JWT verification failed');
+		}
+
+		if (decoded.aud !== this.MCP_AUDIENCE) {
+			throw new Error(`Invalid token audience: expected ${this.MCP_AUDIENCE}`);
+		}
+
+		const accessTokenRecord = await this.accessTokenRepository.findOne({
+			where: { token },
+		});
+
+		if (!accessTokenRecord) {
+			throw new Error('Invalid access token: not found in database');
+		}
+
+		return {
+			token,
+			clientId: decoded.client_id,
+			scopes: [],
+			extra: {
+				userId: decoded.sub,
+			},
+		};
+	}
+
+	async verifyOAuthToken(token: string): Promise<User | null> {
+		try {
+			const authInfo = await this.verifyAccessToken(token);
+
+			const userId = authInfo.extra?.userId as string;
+			if (!userId) {
+				return null;
+			}
+
+			const user = await this.userRepository.findOne({
+				where: { id: userId },
+				relations: ['role'],
+			});
+
+			return user;
+		} catch (error) {
+			return null;
+		}
+	}
+
+	async revokeAccessToken(token: string, clientId: string): Promise<boolean> {
+		const accessTokenRecord = await this.accessTokenRepository.findOne({
+			where: {
+				token,
+				clientId,
+			},
+		});
+
+		if (!accessTokenRecord) {
+			return false;
+		}
+
+		await this.accessTokenRepository.remove(accessTokenRecord);
```

Comment:
We can also just remove directly and check affected rows, no need to query first